gh-824 Add Spring Restdocs Infrastructure

ghillert opened the pull request:
Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/824
Resolves spring-cloud/spring-cloud-dataflow#821

ericbottard commented on spring-cloud-dataflow-docs/src/main/asciidoc/api-guide.adoc:
```diff
@@ -0,0 +1,113 @@
+
+[partintro]
+--
+In this section you will learn all about Spring Cloud Data Flows REST API.
```
Spring Cloud Data Flow**s**

ericbottard replied:
Moreover, and this goes beyond this simple PR, we should settle on Data Flow (here) or Dataflow (line 15), but not both

markfisher replied:
should be "Data Flow"

ericbottard commented on spring-cloud-dataflow-docs/src/main/asciidoc/api-guide.adoc:
```diff
@@ -0,0 +1,113 @@
+
+Every response has the following header(s):
+
+TBD
```
I understand this is to set up the infrastructure, but should we leave these TBDs here?

ericbottard commented on spring-cloud-dataflow-docs/src/main/asciidoc/api-guide.adoc:
```diff
@@ -0,0 +1,113 @@
+
+==== Request structure
+
+include::{snippets}/apps/http-request.adoc[]
```
The generated snippet looks like
```
GET /apps?type=source&type=source HTTP/1.1
Accept: application/json
Host: localhost:8080
```

See the double "source"

ericbottard commented on spring-cloud-dataflow-docs/src/main/asciidoc/api-guide.adoc:
```diff
@@ -0,0 +1,113 @@
+
+==== Example request
+
+include::{snippets}/apps/curl-request.adoc[]
```
Same problem here, double "type"

ericbottard commented on spring-cloud-dataflow-server-core/pom.xml:
```diff
@@ -160,6 +169,15 @@
+				<artifactId>maven-surefire-plugin</artifactId>
+				<configuration>
+					<includes>
+						<include>**/*Documentation.java</include>
```
This effectively disables all the other tests (the ones in `*Tests`)

ericbottard commented:
Overall, looks good, but I think it would be way more manageable if the tests were part of the `-docs` project.
The REST Docs tests can still be called `*Tests` btw, eg `DocumentationTests`. This would simplify configuration

ghillert commented:
Refactored the infrastructure so that the Rest Docs tests use a "representative" SCDF server instance. Also added more documentation. Should probably move the `api documentation tests` to the docs-modules.

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/AppRegistryDocumentation.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class AppRegistryDocumentation extends BaseDocumentation {
+
+	@ClassRule
```
This can move up

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/AppRegistryDocumentation.java:
```diff
@@ -0,0 +1,59 @@
+	public static TestRule springDataflowServer = localDataflowResource;
+
+	@Rule
+	public JUnitRestDocumentation restDocumentation =
```
This can move up

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/AppRegistryDocumentation.java:
```diff
@@ -0,0 +1,59 @@
+			new JUnitRestDocumentation(TARGET_DIRECTORY);
+
+	@Before
+	public void setupMocks() {
```
This can move up

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/BaseDocumentation.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * @author Gunnar Hillert
+ */
+public class BaseDocumentation {
```
This should be made abstract

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/ApiDocumentation.java:
```diff
@@ -0,0 +1,134 @@
+	public void index() throws Exception {
+		this.mockMvc.perform(get("/"))
+			.andExpect(status().isOk())
+			.andDo(print())
```
Please remove `print()`

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/ApiDocumentation.java:
```diff
@@ -0,0 +1,134 @@
+					linkWithRel("streams/deployments").description("Resource for Stream deployment operations"),
+					linkWithRel("streams/deployments/deployment").description("Request (un-)deployment of an existing stream definition"),
+					linkWithRel("counters").description("Counter resource"),
+					linkWithRel("counters/counter").description(""),
```
Missing descriptions

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/ApiDocumentation.java:
```diff
@@ -0,0 +1,134 @@
+					linkWithRel("tasks/deployments").description("Resource for deployment operations"),
+					linkWithRel("tasks/deployments/deployment").description("Launch a task"),
+					linkWithRel("tasks/executions").description("Returns Task executions"),
+					linkWithRel("tasks/executions/name").description("REturns all task execution for a given Task name"),
```
Typo in REturns

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/ApiDocumentation.java:
```diff
@@ -0,0 +1,134 @@
+					linkWithRel("jobs/executions/execution").description("Details for a specific JobExecution"),
+					linkWithRel("jobs/executions/execution/steps").description("All steps for a JobExecution"),
+					linkWithRel("jobs/executions/execution/steps/step").description("Details for a specific step"),
+					linkWithRel("jobs/executions/execution/steps/step/progress").description("Progress information for  a specific step"),
```
Extra space

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/ApiDocumentation.java:
```diff
@@ -0,0 +1,134 @@
+			.andExpect(status().isOk())
+			.andDo(print())
+			.andDo(this.documentationHandler.document(
+				links(
```
The descriptions in here use different forms: sometimes verbs, sometimes nouns. Sometimes they state the obvious, sometimes they add more context. We should be consistent IMHO

ericbottard commented on spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/BaseDocumentation.java:
```diff
@@ -0,0 +1,59 @@
+				.alwaysDo(this.documentationHandler)
+				.build();
+
+//		AppRegistry appRegistry = localDataflowResource.getWebApplicationContext().getBean(AppRegistry.class);
```
This can be removed

ericbottard commented on spring-cloud-dataflow-docs/src/main/asciidoc/api-guide.adoc:
```diff
@@ -0,0 +1,140 @@
+[[api-guide-resources-index-links]]
+===== Links
+
+The main element of the index are the links as they allow you to traverse the Api
```
Api should be spelled API

ericbottard commented on spring-cloud-dataflow-docs/src/main/asciidoc/api-guide.adoc:
```diff
@@ -0,0 +1,140 @@
+The main element of the index are the links as they allow you to traverse the Api
+and execute the desired functionality:
+
+include::{snippets}/api-documentation/index/links.adoc[]
```
Do we actually want to document all those links here, given they're gonna be the meat of the whole documentation anyway?

ghillert commented:
Moved everything over to the `spring-cloud-dataflow-docs` module. More polishing and tests follow tomorrow.

ericbottard commented:
Awesome!
Let me know when it's ready to review/merge

ericbottard commented on spring-cloud-dataflow-docs/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/ApiDocumentation.java:
```diff
@@ -0,0 +1,127 @@
+ */
+public class ApiDocumentation extends BaseDocumentation {
+
+	@ClassRule
```
All of these can move up to the BaseDocumentation class

ericbottard commented on spring-cloud-dataflow-docs/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/ApiDocumentation.java:
```diff
@@ -0,0 +1,127 @@
+			.andExpect(status().isOk())
+			.andDo(this.documentationHandler.document(
+				links(
+					linkWithRel("apps").description("Handle registered applications"),
```
Some descriptions missing, and the tense of all of them should be harmonized (ie use all nouns, or all verbs, but not both)

ghillert replied:
Should be addressed now.

ericbottard commented on spring-cloud-dataflow-docs/src/test/java/org/springframework/cloud/dataflow/server/rest/documentation/BaseDocumentation.java:
```diff
@@ -0,0 +1,59 @@
+				.alwaysDo(this.documentationHandler)
+				.build();
+
+//		AppRegistry appRegistry = localDataflowResource.getWebApplicationContext().getBean(AppRegistry.class);
```
Can be removed

ericbottard commented on spring-cloud-dataflow-server-local/pom.xml:
```diff
@@ -40,6 +40,17 @@
+				<executions>
+					<execution>
+						<goals>
+							<goal>test-jar</goal>
```
Will this artifact be published you think?

ghillert replied:
It should be.

ericbottard commented:
Squashed and merged as e5db3d9 with some additional cleanup
